### PR TITLE
Add `forceUnwrap` and smart-detection of `unwrap` safety 

### LINKED ISSUE #44: throw an error if you use `unwrap` without checking first if you have a `none` or `some`
I'm not sure this is a good idea or not yet, but what if we keep track per instance if a given Option has been "checked", and if you attempt to `unwrap` before checking we throw an error, even if it is a `some`.

If you are working with ts, this probably isnt as much of a concern - but if you are working from javascript directly, theres nothing to prevent you from misunderstanding the api and not checking first.  If you just happen to only test with values that produce `some`s, you might never realize you forgot a check to make sure thats safe first.

Imagine a newcomer writing code like this:

```
service.methodThatReturnsAPromiseOfAnOption()
   .then(o => o.unwrap())
   .then(doSomethingCool);
```

They're new to `Option`s, they read enough of the api to know that `unwrap` is how they get the value they care about out and they move along.  

The error could say something like 
> You attempted to unwrap an `Option` without first checking to make sure it has a value! see this url for more information, but consider using `unwrapOr` or using `.isSome()` or `.isNone()` to test if it is safe to `unwrap` first!

I'm not certain if there would be edge cases where they _could_ know its safe but we cant know that they know - might be something around where we create new instances in some operations, but would need to think about that more - most of those operations could return null and so the point of creating a new instance is to make that a safe operation - meaning if you did check before (on an earlier instance) that doesn't mean you are still safe now.

Also not certain if there should be a way to say "leave me alone, I know what im doing" - or what it would look like if there were.

### PR DESCRIPTION
**Pull request by iwburns**
resolves #44 

@ryanguill let me know what you think here.  This is just a first swing. I believe I got everything from our discussion, but it's very possible I missed something.

I'd also like to get feedback on the error and warning messages I used.  I feel like they're probably fine, but there may be a better way to express our intent there.

### CODE REVIEW

File: src/OptionT.ts [hunk 1]
```diff
@@ -83,22 +83,69 @@ export default abstract class OptionT<T> {
...
+   * Returns the value contained by this [[OptionT]] if it is a `Some`.  Throws an error
+   * containing `message` if this [[OptionT]] is a `None` or a default `message` if one is
+   * not provided.
+   *
+   * #### Note ####
+   * This function is not recommended for normal/production use.  It is almost always
+   * preferable to use [[OptionT.unwrap]] after first having confirmed that your [[OptionT]]
+   * is a `Some` value.
+   *
+   * This function is provided for the sole purpose of quick prototyping.  As such, this function
```
Comment (ryanguill) [inline src/OptionT.ts]: Maybe lets say this is a utility function and not prescribe specifically what it should be used for.

File: src/OptionT.ts [hunk 2]
```diff
@@ -353,6 +405,17 @@ class Some<T> extends OptionT<T> {
   }
 
   unwrap(message?: string): T {
+    if (!this.hasBeenInspected) {
+      throw new Error(
+        'Called unwrap without first checking if it was safe to do so. Please verify' +
+          ' that the OptionT in question is a `Some` value before calling this function.'
```
Comment (ryanguill) [inline src/OptionT.ts]: "... or use one of the other methods that provide safe defaults, like `unwrapOr`" - something like that - I think in general we should encourage people to not use `unwrap` where possible just because of this need to check first if it is safe.

File: src/OptionT.ts [hunk 3]
```diff
@@ -353,6 +405,17 @@ class Some<T> extends OptionT<T> {
...
+      );
+    }
+    return this.value;
+  }
+
+  forceUnwrap(message?: string): T {
+    console.warn('Warning: called forceUnwrap on a `Some` value.  This is not recommended usage.');
```
Comment (ryanguill) [inline src/OptionT.ts]: We might should add error/warning codes to our messages - something ive seen other projects do is actually make their codes urls to documentation, so its easy to google but also to get more information about the problem and possible solutions.

I think steering them to use `unwrap` or `unwrapOr` might be good here too.

File: src/OptionT.ts (hunk 1 above)
Comment (iwburns) [reply src/OptionT.ts]: agree here, the language I used is probably too strong

File: src/OptionT.ts (hunk 2 above)
Comment (iwburns) [reply src/OptionT.ts]: agree, I'll re-word here as well

File: src/OptionT.ts (hunk 3 above)
Comment (iwburns) [reply src/OptionT.ts]: >We might should add error/warning codes to our messages[...]

This is an interesting idea.  I'd like to see an example of this if you have one.  Is this similar to what Rust does with their compiler error codes?  If you're not familiar they basically have a giant error page with further explanations for all of the error codes in their docs somewhere.  I never get there other than by googling an error code, so I'm not exactly sure where they are.

Only concern here is that those URL links are something extra we'd need to maintain if the docs location ever changed.  If this is going to end up being a slightly more involved changed then maybe we should make an issue about it and approach it in a separate PR?  Not sure there.

>I think steering them to use `unwrap` or `unwrapOr` might be good here too.

Agree, will make this change as well.

Comment (ryanguill) [reply src/OptionT.ts]: I think for now lets just create a specific error code that we can include in documentation - something like `nullshield:force_unwrap_warning: called forceUnwrap on ...` so that its easy to copy and paste into google.

File: src/OptionT.ts [hunk 4]
```diff
@@ -113,6 +109,20 @@ export default abstract class OptionT<T> {
...
    * @returns {T}
+   * @throws a `nullshield:unchecked_unwrap` error if you attempt to call it before first checking
+   * if the [[OptionT]] is a safe to unwrap (in other words, it must be a `Some` value).
+   * @throws a `nullshield:unwrap_on_none` error if you attempt to call it on a `None` value after
+   * first checking that the [[OptionT]] is safe to unwrap.
+   *
+   * ## `nullshield:unchecked_unwrap:` ##
+   * To avoid this issue, either check that the given [[OptionT]] is a `Some` value (with
+   * [[OptionT.isSome]] or [[OptionT.isNone]]) or use [[OptionT.unwrapOr]] instead which allows you
+   * to specify a default value in the case where the given [[OptionT]] is a `None`.
```
Comment (ryanguill) [inline src/OptionT.ts]: How about 
> ... or instead use a different function like [[OptionT.unwrapOr]] which allows you to specify a default value, handling the case where the given [[OptionT]] is a `None`.

I like pushing them to `unwrapOr` - its the quick and easy solution to their current problem - but I kinda want to hint that there are other functions as well that could be used to be safer and would maybe express their problems and intent better.

Comment (iwburns) [reply src/OptionT.ts]: I agree.  And since this is in the documentation we've got more room for explanation.

how is something like this:
```
The most direct way to avoid this issue is to either check that the given [[OptionT]] is a `Some` value (with [[OptionT.isSome]] or [[OptionT.isNone]]) or use [[OptionT.unwrapOr]] instead which allows you to specify a default value in the case where the given [[OptionT]] is a `None`.  

However, oftentimes you may not want to simply get the value out of the [[OptionT]]; instead you may want to conditionally use that value in some sort of computation.  In those cases it's likely more clean/clear to use [[OptionT.map]] or a similar function instead.
```

Comment (iwburns) [reply src/OptionT.ts]: @ryanguill no rush, just wanted to make sure you got a notification about this.

Comment (ryanguill) [reply src/OptionT.ts]: LGTM

**Comment by iwburns** [comment]
@ryanguill I'm done with the changes here.  One last check that you're good with this and I'll merge.

Thanks!

**Comment by iwburns** [comment]
went ahead and merged b/c I just realized that you had already approved

**Comment by iwburns** [comment]
@ryanguill v0.8.0 is up on npm. just fyi
